src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ProducerConsumerStream.cs

```diff
@@ -0,0 +1,350 @@
+                }
+
+                _dataAvail.TrySetResult(bytesWritten);
+            }
+        }
+
+        public override bool CanRead
+        {
+            get { return !_disposed; }
+        }
+
+        public override bool CanSeek
+        {
+            get { return false; }
+        }
+
+        public override bool CanWrite
+        {
+            get { return !_disposed; }
+        }
+
+        public override long Length
+        {
+            get { throw new NotImplementedException(); }
+        }
+
+        public override long Position
+        {
+            get { throw new NotImplementedException(); }
+            set { throw new NotImplementedException(); }
+        }
+
+        public override async Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException("buffer");
+            }
+            if (offset < 0 || offset > buffer.Length)
+            {
+                throw new ArgumentOutOfRangeException("offset");
+            }
+            if (count < 0 || count > buffer.Length - offset)
+            {
+                throw new ArgumentOutOfRangeException("count");
+            }
+            Contract.EndContractBlock();
+
+            if (_disposed)
+            {
+                throw new ObjectDisposedException("ProducerConsumerStream");
+            }
+
+            while (count > 0)
+            {
+                if (_currentBuffer == WriteBufferWrapper.EmptyContainer)
+                {
+                    _currentBuffer = await _buffer.Task;
+                }
+
+                int bytesWritten = _currentBuffer.Write(buffer, offset, count);
+                count -= bytesWritten;
+                offset += bytesWritten;
+                if (_currentBuffer.Count == 0)
+                {
+                    _currentBuffer = WriteBufferWrapper.EmptyContainer;
+                }
+
+                _dataAvail.TrySetResult(bytesWritten);
+            }
+        }
+
+        public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException("buffer");
+            }
+            if (offset < 0 || offset > buffer.Length)
+            {
+                throw new ArgumentOutOfRangeException("offset");
+            }
+            if (count < 0 || count > buffer.Length - offset)
+            {
+                throw new ArgumentOutOfRangeException("count");
+            }
+            Contract.EndContractBlock();
+
+            if (_disposed)
+            {
+                return 0;
+            }
+
+            _buffer.TrySetResult(new WriteBufferWrapper(buffer, offset, count));
+            int totalBytesRead = 0;
+            do
+            {
+                int bytesRead = await _dataAvail.Task;
```

Comment:
That is actually a very difficult thing to get right. If the read cancellation token is cancelled, there needs to be a mechanism to prevent any more writing into the current pending buffer. So I would need to track when a write begins and ends so that the read doesn't throw from the cancellation while the buffer continues to be written to. Then there needs to be a mechanism to replace the current buffer with WriteBufferWrapper.EmptyContainer in a thread safe cache coherent manner so that another incoming write doesn't trample over the cancelling read.  We also have the problem of a read or a write could be asynchronously waiting for the corresponding opposite operation to happen which never comes so we need a way to cancel the await on the passed in cancellation token. Checking the code which will use this, it doesn't look like this will ever be cancelled from the read side as this code path will only be used on .Net Native.